### PR TITLE
Fix picking and highlighting regressions.

### PR DESCRIPTION
**Pull request by 1chandu**
Fix for #979

Verified picking and highlighting with `layer-browser` and running website (examples).

### CODE REVIEW

File: src/core/lib/draw-layers.js [hunk 1]
```diff
@@ -197,18 +197,18 @@ function drawLayersInViewport(gl, {
       );
 
       // Blend parameters must not be overriden
-      parameters = Object.assign({viewport: glViewport}, layer.props.parameters || {}, parameters);
+      let layerParameters = Object.assign({viewport: glViewport}, layer.props.parameters || {});
```
Comment (1chandu) [inline src/core/lib/draw-layers.js]: @ibgreen not using previous layer's params (`parameters` -> `let layerParameters`) didn't fix this issue but made this change based on your comment in the issue.

File: src/core/lib/draw-layers.js [hunk 2]
```diff
@@ -197,18 +197,18 @@ function drawLayersInViewport(gl, {
...
 
       if (drawPickingColors) {
-        parameters = Object.assign(parameters, {
+        layerParameters = Object.assign(layerParameters, {
```
Comment (ibgreen) [inline src/core/lib/draw-layers.js]: Let's drop the assignment, Object.assign modifies its first argument.
```
Object.assign(layerParameters, {
```

File: src/core/lib/draw-layers.js [hunk 3]
```diff
@@ -239,11 +239,11 @@ function updateLayerHighlightColor(layer) {
...
 
-  if (layer.state.model) {
-    const pickingSelectedColorValid = layer.props.highlightedObjectIndex >= 0;
-    const pickingSelectedColor = pickingSelectedColorValid ?
-      layer.encodePickingColor(layer.props.highlightedObjectIndex) :
-      layer.nullPickingColor();
+  // Update picking module settings if highlightedObjectIndex is set.
+  // This will overwrite any settings from auto highlighting.
+  const pickingSelectedColorValid = layer.props.highlightedObjectIndex >= 0;
+  if (layer.state.model && pickingSelectedColorValid) {
```
Comment (ibgreen) [inline src/core/lib/draw-layers.js]: Concern: What happens when we transition from valid to invalid? Won't the module settings need to be updated? This change seems to restrict updates to the valid case?

File: src/core/lib/layer-manager.js [hunk 4]
```diff
@@ -658,8 +658,8 @@ export default class LayerManager {
     const radius = this._pickingRadius;
     const selectedInfos = this.pickObject({x: pos.x, y: pos.y, radius, mode: 'hover'});
 
-    const firstInfo = selectedInfos.find(info => info.index >= 0);
-    if (firstInfo && this._onLayerHover) {
+    const firstInfo = selectedInfos.find(info => info.index >= 0) || null;
```
Comment (ibgreen) [inline src/core/lib/layer-manager.js]: Maybe a comment about the `null`, 
```
// Note: API docs state that info must be `null`. 
```

And why is this only in the hover case, not in the click case?

Share code between the two cases?

File: src/core/lib/draw-layers.js (hunk 1 above)
Comment (ibgreen) [reply src/core/lib/draw-layers.js]: > not using previous layer's params (parameters -> let layerParameters) didn't fix this issue but made this change based on your comment in the issue.

Yes good. Perhaps there is one more bug, but your change here is correct.

File: src/core/lib/draw-layers.js (hunk 3 above)
Comment (1chandu) [reply src/core/lib/draw-layers.js]: During picking cycle, we update `pickingSelectedColorValid` to true/false based on valid/invalid transition. Here we overwrite that if `highlightedObjectIndex` is set.

**Pull request merged**